LibraBFT/Impl/Consensus/RoundManager/Properties.agda

```diff
@@ -0,0 +1,190 @@
+    pf (Left _) st outs (constructAndSignVoteMSpec.mkContract noOuts (mkNoEpochChange es≡₁ es≡₂) (mkNoVoteCorrect lv≡ lvr≤)) =
+      mkContract noOuts (mkNoEpochChange es≡₁ es≡₂) (inj₁ (mkNoVoteCorrect lv≡ lvr≤))
+    pf (Right vote) st outs (constructAndSignVoteMSpec.mkContract noOuts nec vc) ._ refl =
+      PersistentLivenessStorageProps.saveVoteMSpec.contract vote
+        (RWST-weakestPre-ebindPost unit (λ _ → ok vote) Contract-++outs) st
+        (mkContract
+          (++-NoMsgOuts outs [] noOuts refl)
+          (transNoEpochChange noEpochChange-pre-preUpdated nec)
+          (Right voteNotSaved))
+        λ where
+          _ .unit refl →
+            mkContract (++-NoMsgOuts outs [] noOuts refl)
+              (transNoEpochChange (noEpochChange-pre-preUpdated) (transNoEpochChange nec (mkNoEpochChange refl refl)))
+              voteSaved
+      where
+      Contract-++outs = RWST-Post++ (Contract pre) outs
+
+      voteNotSaved : VoteNotSaved pre st epoch round
+      voteNotSaved = vote , substVoteCorrect refl refl refl refl eb≡b-epoch eb≡b-round vc
+
+      voteSaved : ∀ {bs} → VoteCorrect pre (rmSetBlockStore st bs) epoch round vote
+      voteSaved = substVoteCorrect refl refl refl refl eb≡b-epoch eb≡b-round vc
+
+  contract
+    : ∀ pre Post
+      → (∀ r st outs → Contract pre r st outs → Post r st outs)
+      → LBFT-weakestPre (executeAndVoteM b) Post pre
+  contract pre Post pf =
+    RWST-⇒ (Contract pre) Post pf (executeAndVoteM b) unit pre (contract' pre)
+
+module processProposalMSpec (proposal : Block) where
+  open import LibraBFT.Impl.Consensus.Liveness.Properties.ProposerElection
+  open import LibraBFT.Impl.Consensus.BlockStorage.Properties.BlockStore
+  open        LibraBFT.Impl.Consensus.RoundManager.processProposalM proposal
+
+  epoch = proposal ^∙ bEpoch
+  round = proposal ^∙ bRound
+
+  record NoMsgOutsCorrect (pre post : RoundManager) (outs : List Output) : Set where
+    constructor mkNoOutsCorrect
+    field
+      noMsgOuts : NoMsgOuts outs
+      nvc⊎vns   : NoVoteCorrect pre post ⊎ VoteNotSaved pre post epoch round
+
+  record VoteMsgOutsCorrect (pre post : RoundManager) (outs : List Output)  : Set where
+    constructor mkVoteMsgOutsCorrect
+    field
+      vm          : VoteMsg
+      pid         : Author
+      voteMsgOuts : VoteMsgOuts outs vm (pid ∷ [])
+      outCorrect  : VoteCorrect pre post epoch round (vm ^∙ vmVote)
+
+  OutsCorrect : (pre post : RoundManager) (outs : List Output) → Set
+  OutsCorrect pre post outs = NoMsgOutsCorrect pre post outs ⊎ VoteMsgOutsCorrect pre post outs
+
+  record Contract (pre : RoundManager) (_ : Unit) (post : RoundManager) (outs : List Output) : Set where
+    constructor mkContract
+    field
+      noOuts        : OutsCorrect pre post outs
+      inv           : NoEpochChange pre post
+
+  contract : ∀ pre → LBFT-weakestPre (processProposalM proposal) (Contract pre) pre
```

Comment:
I'm still reserving judgment on how monolithic is too monolithic.  I think this proof is on the border, and if it were more complicated than just dismissing a bunch of simple cases and invoking another contract, I'd be inclined to add a comment that it needs to be broken into smaller pieces.  This is just commentary, not asking you to do anything at this stage.